awd/cli.py

```diff
@@ -76,51 +77,67 @@ def deposit(
             e.response["Error"]["Message"],
         )
         return  # Unable to access S3 bucket, exit application
+
     DoiProcessAttempt.set_table_name(config.DOI_TABLE)
     if not DoiProcessAttempt.exists():
         logger.exception("Unable to read DynamoDB table")
         return  # exit application
 
+    unprocessed_dois = DoiProcessAttempt.retrieve_unprocessed_dois()
+
     for doi_file in s3_client.retrieve_file_type_from_bucket(
         config.BUCKET, ".csv", "archived"
     ):
-        dois = get_dois_from_spreadsheet(f"s3://{config.BUCKET}/{doi_file}")
-        for doi in dois:
-            logger.error(doi)
-            article = Article(
-                doi=doi,
-                metadata_url=config.METADATA_URL,
-                content_url=config.CONTENT_URL,
-                s3_client=s3_client,
-                bucket=config.BUCKET,
-                sqs_client=sqs_client,
-                sqs_base_url=config.SQS_BASE_URL,
-                sqs_input_queue=config.SQS_INPUT_QUEUE,
-                sqs_output_queue=config.SQS_OUTPUT_QUEUE,
-                collection_handle=config.COLLECTION_HANDLE,
-            )
-            try:
-                article.process()
-            except (
-                ClientError,
-                GetError,
-                InvalidArticleContentResponseError,
-                InvalidCrossrefMetadataError,
-                InvalidDSpaceMetadataError,
-                UnprocessedStatusFalseError,
-            ):
-                logger.exception(article.doi)
-                continue
+        for doi in [
+            doi
+            for doi in get_dois_from_spreadsheet(f"s3://{config.BUCKET}/{doi_file}")
+            if doi not in unprocessed_dois
+        ]:
+            DoiProcessAttempt.check_doi_and_add_to_table(doi)
+            unprocessed_dois.append(doi)
+
```

Comment:
I might suggest slightly reworking this.  At a glance, it's not obvious that the end result is a list of unique DOIs.

Perhaps a `set()` could be a good option here?  Something along the lines of:

```python
unprocessed_dois = set()

# add unprocessed DOIs from Dynamo
unprocessed_dois.update(DoiProcessAttempt.retrieve_unprocessed_dois())

# add new DOIs from spreadsheets
for doi_file in ...:
    ...
    ...
    unprocessed_dois.add(doi)

# convert to list (and technically, this is optional, as you could iterate through the set as well later)
unprocessed_dois = list(unprocessed_dois)
```

It's not dramatically different, but it allows skipping "does this DOI exist in the list yet?" check when looping through the spreadsheets.

Reply:
Never used this before, great tip, thanks!